torch/csrc/jit/fuser/cpu/fused_kernel.cpp

```diff
@@ -27,6 +53,96 @@ static bool programExists(const std::string& program) {
   return (system(cmd.c_str()) == 0);
 }
 
+#ifdef _MSC_VER
+std::string exec(const std::string& cmd) {
+  std::array<char, 128> buffer;
+  std::string result;
+  std::unique_ptr<FILE, decltype(&_pclose)> pipe(
+      _popen(cmd.c_str(), "r"), _pclose);
+  if (!pipe) {
+    throw std::runtime_error("popen() failed!");
+  }
+  while (fgets(buffer.data(), static_cast<int>(buffer.size()), pipe.get()) != nullptr) {
+    result += buffer.data();
+  }
+  return result;
+}
+
+inline std::string& rtrim(std::string& s, const char* t = " \t\n\r\f\v") {
+  s.erase(s.find_last_not_of(t) + 1);
+  return s;
+}
+
+void activate() {
+  char* root = nullptr;
+  std::string cmd;
+  std::string path;
+  std::string vcruntime_plat;
+  std::string envvars;
+
+  // Checking whether the environment is already activated
+  if (getenv("VSCMD_ARG_TGT_ARCH")) {
+    return;
+  }
+
+  // Getting `ProgramFiles` through environment variable queries
+  root = getenv("ProgramFiles(x86)");
+  if (!root) {
+    root = getenv("ProgramFiles");
+  }
+  if (!root) {
+    return;
+  }
+
+  // Getting VS 2017 installation path through `vswhere`
+  cmd = "\"" + std::string(root) +
+      "\\Microsoft Visual Studio\\Installer\\vswhere.exe\""
+      " -latest -prerelease -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -property installationPath";
+  std::cout << cmd << std::endl;
+  try {
+    path = exec(cmd);
+    rtrim(path);
+  } catch (std::runtime_error&) {
+    return;
+  }
+
+  // Checking whether the activation script `vcvarsall.bat` exists
+  path += "\\VC\\Auxiliary\\Build";
+  struct stat st;
+  if (stat(path.c_str(), &st) == -1 || !(st.st_mode & _S_IFDIR)) {
+    return;
+  }
+  path += "\\vcvarsall.bat";
+  if (_access(path.c_str(), 0) == -1) {
+    return;
+  }
+
+  // Determining current platform
+  if (sizeof(void*) == 8) {
+    vcruntime_plat = "x64";
+  } else {
+    vcruntime_plat = "x86";
+  }
+
+  // Getting environment variables after activating VS development shell
+  cmd = "\"" + path + "\" " + vcruntime_plat + ">NUL && set";
+  try {
+    envvars = exec(cmd);
+  } catch (std::runtime_error&) {
+    return;
+  }
+
+  // Setting environment variables to the current environment
+  std::istringstream f(envvars);
+  std::string envvar;
+  while (getline(f, envvar, '\n')) {
+    if (_putenv(envvar.c_str()) == -1) {
```

Comment:
Nrrrgh. I'm generally not a fan of in-place mutating the environment variables of the program itself. It's global state, so mutations like this are not thread safe, and they can cause spooky action at a distance if users are expecting environment variables to be the same as what they were at program start. Can we just store the environment variables in an array somewhere, and then do subprocess invocation with them set? (I'm not sure how easy or hard the Windows API makes this for you.)

Reply:
An alternative is to store the whole envvar as an intermediate variable and then prepend them when calling `system`, that is `system("SET A=B && SET B=C && ... && cl ....")`, what do you think?

Reply:
That seems fine to me, if system doesn't let you feed it env vars otherwise. I hope we don't hit the command line size limit.

Reply:
I choose `_spawnve` since it has an argument for env vars.